custom/icds/utils/ccz_hosting.py

```diff
@@ -0,0 +1,62 @@
+from __future__ import absolute_import
+from __future__ import unicode_literals
+
+from django.urls import reverse
+from django.utils.functional import cached_property
+
+from corehq.blobs import (
+    get_blob_db,
+    CODES,
+)
+from corehq.blobs.models import BlobMeta
+
+
+class CCZHostingUtility:
+    def __init__(self, ccz_hosting=None, blob_id=None):
+        self.ccz_hosting = ccz_hosting
+        self.blob_id = blob_id
+        if ccz_hosting and not blob_id:
+            self.blob_id = self.ccz_hosting.blob_id
+
+    def file_exists(self):
+        return get_blob_db().exists(key=self.blob_id)
+
+    def get_file(self):
+        return get_blob_db().get(key=self.blob_id)
+
+    def get_file_size(self):
+        return get_blob_db().size(key=self.blob_id)
+
+    @cached_property
+    def get_file_meta(self):
+        if self.file_exists():
+            return get_blob_db().metadb.get(key=self.blob_id, parent_id='CCZHosting')
+
+    @property
+    def ccz_details(self):
+        file_name = self.ccz_hosting.file_name
+        if not file_name and self.get_file_meta:
+            file_name = self.get_file_meta.name
+        return {
+            'name': file_name,
+            'download_url': reverse('ccz_hosting_download_ccz', args=[
+                self.ccz_hosting.domain, self.ccz_hosting.id, self.blob_id])
+        }
+
+    def store_file_in_blobdb(self, ccz_file, name=None):
+        db = get_blob_db()
+        try:
+            kw = {"meta": db.metadb.get(parent_id='CCZHosting', key=self.blob_id)}
```

Comment:
Maybe I suggested too many improvements in one comment. However, I think elimination of `settings.CCZ_FILE_HOSTING_SUPPORTING_FILES` should not be dismissed so lightly:

    # Storing the blob_id and domain in CCZHosting eliminates the need for
    # settings.CCZ_FILE_HOSTING_SUPPORTING_FILES. All of the information
    # that was stored in settings is now available in the CCZHosting table.

I still think it would be good to add a `domain` field to `CCZHosting` so that those `settings` values can be eliminated (@snopoke also [requested something similar](https://github.com/dimagi/commcare-hq/pull/24129/files#r281156258)).

---

You've pointed out in an offline chat that the `blob_id` is a derived field:

https://github.com/dimagi/commcare-hq/blob/5f113267e72f0bf735bf626d3fc68a607e838b78/custom/icds/models.py#L65-L69

and that key uniquely identifies a given blob, and the content of that blob will never change (blob ids are not reused). If all that is true, then it seems reasonable to use a composite/calculated `blob_id` as defined. Additionally, if all that is true then the code of the method on which we are commenting is more complex than necessary:

```py
db.metadb.get(parent_id='CCZHosting', key=self.blob_id)
```

will always raise `BlobMeta.DoesNotExist` (blob ids are not reused), and therefore it would be better to remove the `try:` block and change this method to do:

```py
db.put(
    ccz_file,
    domain=self.ccz_hosting.link.domain,
    # varying parent_id means the metadata records will be distributed
    # across all shards of the metadata database.
    parent_id="CCZHosting-%s" % self.ccz_hosting.id,
    type_code=CODES.tempfile,
    key=self.blob_id,
    name=name,
)
```

If an error is raised due to an existing blob that would be overwritten then there is an error in the code or in your assumptions about the uniqueness of `blob_id`, and we want to know about that.

Reply:
@millerdev 
If i have the same version of "app_id + version + profile" for two links, this would fail when i would try to save blobmeta.
Even if i was to get pass that, in task `setup_ccz_file_for_hosting` that does the setup, i would always need to generate the ccz to pass to this method.

Working on removing things from settings as CCZHosting objects.
One thing i would need to do is to either
1. remove the validations on model for app_id and version since it would need to be blank for the supporting apps and docs.
2. or set up value like "dummy" for app_id and 0 for version for these docs. [I would prefer this one because i would like to keep the validations on the model]

Reply:
Update:
I added a new model instead for supporting files instead of making this complex.

Reply:
I'm not sure how to convince you anymore. If line 48 does not raise `BlobMeta.DoesNotExist`

https://github.com/dimagi/commcare-hq/blob/e2c9f5ad2b0f4e1260ea5e4f5389fb79b55aa19d/custom/icds/utils/ccz_hosting.py#L47-L49

 then this code is reusing the blob id (that is, writing new content to an existing blob key), which you previously said will never happen.

Reply:
I hear you.
>However, I think elimination of settings.CCZ_FILE_HOSTING_SUPPORTING_FILES should not be dismissed so lightly

i made this change.

> will always raise BlobMeta.DoesNotExist (blob ids are not reused)

I see that check is confusing, removed it [here](https://github.com/dimagi/commcare-hq/pull/24129/commits/7155265a6b6f46fc09657052dc4552fb37dc7861) since we would not restore an already present blob

Reiterating on the why i cant use "CCZHosting-%s" % self.ccz_hosting.id" with the same blob id since that is not allowed due to unique constraint on blob id(key) for blob meta